File: deps/src/inbreeding.c
#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>

/**
 * Create the inbreeding coefficients according to the algorithm given
 * in "Comparison of four direct algorithms for computing inbreeding
 * coefficients" by Mehdi Sargolzaei and Hiroaki Iwaisaki, Animal
 * Science Journal (2005) 76, 401--406.  This function is a modified
 * version of the code published in an appendix to that paper.
 *
 * @param sire 1-based indices of animal's sire (0 => unknown)
 * @param dam  1-based indices of animal's dam (0 => unknown)
 * @param lap  longest ancestral path.  lap[0] == -1, for unknown parent
 * @param n    length of sire and dam.  lap and F are of length n+1
 * @param F    double vector of length n that contains the inbreeding coefficients on return
 *
 */

void printivec(const ptrdiff_t* vv, const char* nm, ptrdiff_t l) {
    ptrdiff_t i;
    printf("%s = [%ld", nm, vv[0]);
    for(i = 1; i < l; ++i) printf(", %ld", vv[i]);
    printf("]\n");
}

void inbreeding(const ptrdiff_t* sire, const ptrdiff_t* dam, const ptrdiff_t *lap,
		ptrdiff_t n, double *F) {
    ptrdiff_t i, j, t, maxlap, S, D;
    ptrdiff_t *SI, *MI;		/* start and minor */
    double *L = calloc(n + 1, sizeof(double)), *B = calloc(n + 1, sizeof(double));
    ptrdiff_t *Anc = calloc(n + 1, sizeof(ptrdiff_t));	/* ancestor */
	
    maxlap = lap[0];	     /* determine the maximum ancestral path length */
    for (i = 1; i <= n; ++i) {
	if (lap[i] > maxlap) maxlap = lap[i];
    }
    SI = calloc(maxlap + 1, sizeof(ptrdiff_t)); /* initialized to zeros */
    MI = calloc(maxlap + 1, sizeof(ptrdiff_t));

    F[0] = -1.0;		     /* set F  for unknown parents */
    for(i = 1; i <= n; i++) {	     /* evaluate F */
	S = sire[i-1]; D = dam[i-1]; /* parents of animal i */
	printf("(i,S,D) => (%ld,%ld,%ld)\n",i,S,D);
	B[i] = 0.5 - 0.25 * (F[S] + F[D]); 
				/* adjust start and minor */
	for (j = 0; j < lap[i]; j++) {++SI[j]; ++MI[j];} 
	printivec(SI,"SI",maxlap);
	printivec(MI,"MI",maxlap);
	if (S == 0 || D == 0) { /* both parents unknown */
	    printf("unknown parents exit\n");
	    F[i] = L[i] = 0.0; 
	    continue;
	}
	if(S == sire[i-2] && D == dam[i-2]) { /* full-sib with last animal */
	    printf("full sib exit\n");
	    F[i] = F[i-1];
	    L[i] = L[i-1];
	    continue;
	}
    	F[i] = -1.0; L[i] = 1.0; 
	t = lap[i]; /* largest lap group number in the animal's pedigree */
	Anc[MI[t]++] = i; /* initialize Anc and increment MI[t] */
	while(t > -1) { /* from the largest lap group to zero */
	    j = Anc[--MI[t]]; /* next ancestor */
	    printf("(i,j,t) => (%ld,%ld,%ld)\n",i,j,t);
	    printivec(SI,"SI",maxlap);
	    printivec(MI,"MI",maxlap);
	    printivec(Anc,"Anc",n+1);
	    S = sire[j-1]; D = dam[j-1]; /* parents of the ancestor */
	    if (S) {
		if (!L[S]) Anc[MI[lap[S]]++] = S; 
				/* add sire in its lap group in Anc
				 * array if it is not added yet and
				 * increment the minor index for the group */ 
		L[S] += 0.5 * L[j]; /* contribution to sire */
	    }
	    if (D) {
		if (!L[D]) Anc[MI[lap[D]]++] = D;
		L[D] += 0.5 * L[j]; /* contribution to dam */
	    }
	    F[i] += L[j] * L[j] * B[j];
	    L[j] = 0; /*clear L[j] for the evaluation of the next animal */
	    if (MI[t] == SI[t]) --t; /* move to the next lap group when
				      * all ancestors in group t have been
				      * evaluated */
	} 
    }
    free(L); free(B); free(Anc); free(SI); free(MI);
}

File: src/inbreeding.jl
# evaluate the inbreeding coefficients in a pedigree
function inbreeding(p::Pedigree)
    sire = p.sire
    dam = p.dam
    lap = p.lap
    n = length(sire)
    ## Arrays of length n+1 have a placeholder for missing parents in the first element
    F = zeros(n+1)          # inbreeding coefficients
    ccall((:inbreeding,libpedigree),Void,(Ptr{Int},Ptr{Int},Ptr{Int},Int,Ptr{Cdouble}),
          sire,dam,lap,n,F)
    return F
    
    F[1] = -1.                      # unknown parent
    L = zeros(n+1)
    B = zeros(n+1)                  # diagonal factor in A
    Anc = zeros(Int,n+1)            # Ancestor
    maxlap = maximum(lap)
    SI = zeros(Int,maxlap+1)        # start indices
    MI = zeros(Int,maxlap+1)        # minor indices
    for i in 1:n
        S = sire[i]
        D = dam[i]
        B[i] = 0.5 - 0.25*(F[S+1]+F[D+1])
        for j in 1:lap[i+1]         # adjust start and minor
            SI[j+1] += 1
            MI[j+1] += 1
        end
        @show SI
        @show MI
        if S == 0 && D == 0         # both parents unknown
            F[i+1] = L[i+1] = 0.    # non-inbred
            continue
        end
        if i > 1 && S == sire[i-1] && D == dam[i-1]
            F[i+1] = F[i]           # full sib with last animal
            L[i+1] = L[i]
            continue
        end
        F[i+1] = -1.
        L[i+1] = 1.
        t = lap[i+1]       # longest ancestral path length
        @show i,S,D,t
        @show F
        @show L
        @show SI
        @show MI
        Anc[MI[t+1]+1] = i   # initialize Ancestors with this animal
        MI[t+1] += 1 # increment to compensate for decrement below
        while t ≥ 0
            @show t, MI[t+1]
            MI[t+1] -= 1
            j = Anc[MI[t+1]+1]   # next ancestor
            @show i,j,t
            S = sire[j]             # parents of the ancestor
            @show S,lap[S+1]
            if S > 0
                if (L[S+1] != 0)
                    @show MI[lap[S+1]+1]
                    Anc[MI[lap[S+1]+1]+1] = S # add sire to Ancestor array
                    MI[lap[S+1]+1] += 1     # increment minor index for the group
                end
                L[S+1] += 0.5 * L[j+1]
            end
            D = dam[j]
            @show D,lap[D+1]
            if D > 0                # same for dam
                if (L[D+1] != 0)
                    Anc[MI[lap[D+1]+1]+1] = D
                    MI[lap[D+1]+1] += 1
                end
                L[D+1] += 0.5 * L[j+1]
            end
            @show B
            @show L
            @show Anc
            F[i+1] += abs2(L[j+1]) * B[j]
            @show F
            L[j+1] = 0. # clear L[j] for evaluation of next animal
            if MI[t+1] == SI[t+1]           # move to the next LAP group
                t -= 1
            end
        end
        end
    F .+ 1.
end
